File: tdd-contract-review/benchmark/sample-app/app/services/transaction_service.rb
# frozen_string_literal: true

class TransactionService
  Result = Struct.new(:success?, :transaction, :error, :details, keyword_init: true)

  def initialize(user:, wallet:, params:)
    @user = user
    @wallet = wallet
    @params = params
  end

  def call
    validate_wallet_active!
    validate_currency_match!
    validate_sufficient_balance!

    transaction = build_transaction
    transaction.save!

    deduct_balance!(transaction)

    if transaction.payment?
      charge_payment_gateway(transaction)
    end

    Result.new(success?: true, transaction: transaction)
  rescue ActiveRecord::RecordInvalid => e
    Result.new(success?: false, error: 'Validation failed', details: e.record.errors.full_messages)
  rescue WalletInactiveError => e
    Result.new(success?: false, error: e.message, details: ['Wallet must be active'])
  rescue CurrencyMismatchError => e
    Result.new(success?: false, error: e.message, details: ['Currency must match wallet currency'])
  rescue InsufficientBalanceError => e
    Result.new(success?: false, error: e.message,
               details: ["Current balance: #{@wallet.balance}, requested: #{@params[:amount]}"])
  rescue PaymentGateway::ChargeError => e
    Result.new(success?: false, error: 'Payment processing failed', details: [e.message])
  end

  private

  def validate_wallet_active!
    raise WalletInactiveError, 'Wallet is not active' unless @wallet.active?
  end

  def validate_currency_match!
    return if @params[:currency] == @wallet.currency

    raise CurrencyMismatchError, 'Currency does not match wallet'
  end

  def validate_sufficient_balance!
    amount = BigDecimal(@params[:amount].to_s)
    return if @wallet.balance >= amount

    raise InsufficientBalanceError,
          "Insufficient balance (need #{amount}, have #{@wallet.balance})"
  end

  def deduct_balance!(transaction)
    return if transaction.deposit?

    @wallet.withdraw!(transaction.amount)
  end

  def build_transaction
    @user.transactions.build(
      wallet: @wallet,
      amount: @params[:amount],
      currency: @params[:currency],
      description: @params[:description],
      category: @params[:category] || 'transfer',
      status: 'pending'
    )
  end

  def charge_payment_gateway(transaction)
    response = PaymentGateway.charge(
      amount: transaction.amount,
      currency: transaction.currency,
      user_id: @user.id,
      transaction_id: transaction.id
    )

    if response.success?
      transaction.update!(status: 'completed')
    else
      transaction.update!(status: 'failed')
    end
  end

  class WalletInactiveError < StandardError; end
  class CurrencyMismatchError < StandardError; end
  class InsufficientBalanceError < StandardError; end
end

File: tdd-contract-review/benchmark/sample-app-go/internal/service/transaction_service.go
package service

import (
	"errors"
	"fmt"
	"math/big"

	"github.com/example/sample-app-go/internal/model"
	"github.com/example/sample-app-go/internal/payment"
	"github.com/example/sample-app-go/internal/store"
)

type CreateParams struct {
	Amount      *big.Float
	Currency    string
	Description string
	Category    model.TransactionCategory
}

type Result struct {
	Success     bool
	Transaction *model.Transaction
	Error       string
	Details     []string
}

type TransactionService struct {
	repo    *store.MemoryRepo
	gateway payment.Gateway
}

func NewTransactionService(repo *store.MemoryRepo, gateway payment.Gateway) *TransactionService {
	return &TransactionService{repo: repo, gateway: gateway}
}

var (
	ErrWalletInactive     = errors.New("wallet is not active")
	ErrCurrencyMismatch   = errors.New("currency does not match wallet")
	ErrInsufficientFunds  = errors.New("insufficient balance")
	ErrPaymentProcessing  = errors.New("payment processing failed")
)

func (s *TransactionService) Call(user *model.User, wallet *model.Wallet, p CreateParams) Result {
	if err := s.validateWalletActive(wallet); err != nil {
		return failure(err.Error(), []string{"Wallet must be active"})
	}
	if err := s.validateCurrencyMatch(wallet, p); err != nil {
		return failure(err.Error(), []string{"Currency must match wallet currency"})
	}
	if err := s.validateSufficientBalance(wallet, p); err != nil {
		return failure(err.Error(), []string{
			fmt.Sprintf("Current balance: %s, requested: %s", wallet.Balance.Text('f', 8), p.Amount.Text('f', 8)),
		})
	}

	tx := s.buildTransaction(user, wallet, p)
	created, err := s.repo.CreateTransaction(tx)
	if err != nil {
		return failure("Validation failed", []string{err.Error()})
	}

	if err := s.deductBalance(wallet, created); err != nil {
		return failure(err.Error(), []string{err.Error()})
	}

	if created.IsPayment() {
		s.chargePaymentGateway(user, created)
	}

	return Result{Success: true, Transaction: created}
}

func (s *TransactionService) validateWalletActive(w *model.Wallet) error {
	if !w.IsActive() {
		return ErrWalletInactive
	}
	return nil
}

func (s *TransactionService) validateCurrencyMatch(w *model.Wallet, p CreateParams) error {
	if p.Currency != w.Currency {
		return ErrCurrencyMismatch
	}
	return nil
}

func (s *TransactionService) validateSufficientBalance(w *model.Wallet, p CreateParams) error {
	if p.Amount == nil {
		return errors.New("amount is required")
	}
	if w.Balance.Cmp(p.Amount) < 0 {
		return ErrInsufficientFunds
	}
	return nil
}

func (s *TransactionService) buildTransaction(u *model.User, w *model.Wallet, p CreateParams) *model.Transaction {
	cat := p.Category
	if cat == "" {
		cat = model.CategoryTransfer
	}
	return &model.Transaction{
		UserID:      u.ID,
		WalletID:    w.ID,
		Amount:      p.Amount,
		Currency:    p.Currency,
		Description: p.Description,
		Category:    cat,
		Status:      model.TxPending,
	}
}

func (s *TransactionService) deductBalance(w *model.Wallet, tx *model.Transaction) error {
	if tx.IsDeposit() {
		return nil
	}
	return w.Withdraw(tx.Amount)
}

func (s *TransactionService) chargePaymentGateway(u *model.User, tx *model.Transaction) {
	resp, err := s.gateway.Charge(payment.ChargeRequest{
		Amount:        tx.Amount,
		Currency:      tx.Currency,
		UserID:        u.ID,
		TransactionID: tx.ID,
	})
	if err != nil || !resp.Success {
		_ = s.repo.UpdateTransactionStatus(tx.ID, model.TxFailed)
		return
	}
	_ = s.repo.UpdateTransactionStatus(tx.ID, model.TxCompleted)
}

func failure(msg string, details []string) Result {
	return Result{Success: false, Error: msg, Details: details}
}
